Realm/Realm/Handles/SharedRealmHandle.cs

```diff
@@ -454,25 +474,75 @@ public void BeginTransaction()
             nativeException.ThrowIfNecessary();
         }
 
+        public async Task BeginTransactionAsync()
+        {
+            var tcs = new TaskCompletionSource<object>();
+            var tcsHandle = GCHandle.Alloc(tcs);
+            try
+            {
+                _asyncTransactionHandle = NativeMethods.begin_transaction_async(this, GCHandle.ToIntPtr(tcsHandle), out var nativeException);
+                nativeException.ThrowIfNecessary();
+                await tcs.Task;
+            }
+            finally
+            {
+                // The _asyncTransactionHandle is only relevant until the async call is not completed, hence queued.
+                // After it's useless and transactions should just be deleted with the normal NativeMethods.cancel_transaction.
+                _asyncTransactionHandle = null;
+                tcsHandle.Free();
+            }
+        }
+
         public void CommitTransaction()
         {
             NativeMethods.commit_transaction(this, out var nativeException);
             nativeException.ThrowIfNecessary();
         }
 
-        public void CancelTransaction()
+        public async Task CommitTransactionAsync()
         {
-            NativeMethods.cancel_transaction(this, out var nativeException);
-            nativeException.ThrowIfNecessary();
+            var tcs = new TaskCompletionSource<object>();
+            var tcsHandle = GCHandle.Alloc(tcs);
+            try
+            {
+                _asyncTransactionHandle = NativeMethods.commit_transaction_async(this, GCHandle.ToIntPtr(tcsHandle), out var nativeException);
+                nativeException.ThrowIfNecessary();
+                await tcs.Task;
+            }
+            finally
+            {
+                // The _asyncTransactionHandle is only relevant until the async call is not completed, hence queued.
+                // After it's useless and transactions should just be deleted with the normal NativeMethods.cancel_transaction.
+                _asyncTransactionHandle = null;
+                tcsHandle.Free();
+            }
         }
 
-        public bool IsInTransaction()
+        public void CancelTransaction()
         {
-            var result = NativeMethods.is_in_transaction(this, out var nativeException);
+            var nativeException = new NativeException
+            {
+                type = RealmExceptionCodes.NoError
+            };
```

Comment:
`RealmExceptionCodes::NoError = -1`, because of this I thought that's just safer to be explicit about it. This avoids a possible gotcha if in the future `CancelTransaction` is changed and the `nativeException` won't only be an `out` var.

Reply:
I think if we change the signature of `CancelTransaction`, we should update the code then.

Reply:
We never initialize the native exception variable because it always gets overwritten by native. E.g. if you have

```csharp
var somethingFromNative = NativeMethods.invoke_some_method(out var ex);
```

we allocate the memory in managed code, but the fields get set by native. Similarly, if you have an if clause, you can do:

```csharp
NativeException ex;
if (condition)
{
    NativeMethods.call_method_one(out ex);
}
else
{
    NativeMethods.call_method_two(out ex);
}
```